### PR TITLE
[Linalg] Add cusolver syevjBatched path for torch.linalg.eigh when cuda >= 11.3 U1

### PR DESCRIPTION
**Pull request by xwang233**
This PR adds the `cusolverDn<T>SyevjBatched` fuction to the backend of `torch.linalg.eigh` (eigenvalue solver for Hermitian matrix). Using the heuristics from https://github.com/pytorch/pytorch/pull/53040#issuecomment-788264724 and my local tests, the `syevj_batched` path is only used when `batch_size > 1` and `matrix_size <= 32`. This would give us huge performance boost in those cases.

Since there were known numerical issues on cusolver `syevj_batched` before cuda 11.3 update 1, this PR only enables the dispatch when cuda version is no less than that.

See also #42666 #47953 #53040

### CODE REVIEW

**Review by IvanYashchuk — APPROVED** [review]
It's great that the bug is fixed!

File: aten/src/ATen/native/cuda/BatchLinearAlgebraLib.h [hunk 1]
```diff
@@ -21,6 +21,17 @@
   constexpr bool use_cusolver_potrf_batched_ = false;
 #endif
 
+// cusolverDn<T>syevjBatched may have numerical issue before cuda 11.3.1 release,
+// (which is cusolver version 11102 in the header), so we only use cusolver potrf batched
```
[the remark this answers is not in the document]
Comment (xwang233) [reply aten/src/ATen/native/cuda/BatchLinearAlgebraLib.h]: Thanks for finding this! 😂

File: aten/src/ATen/native/cuda/BatchLinearAlgebraLib.cu [hunk 2]
```diff
@@ -1242,16 +1242,21 @@ static void linalg_eigh_cusolver_syevj(const Tensor& eigenvalues, const Tensor&
...
-  // TODO: syevj_batched should be added here, but at least for CUDA 11.2 it contains a bug leading to incorrect results
-  // See https://github.com/pytorch/pytorch/pull/53040#issuecomment-793626268 and https://github.com/cupy/cupy/issues/4847
+static void linalg_eigh_cusolver_syevj_batched(const Tensor& eigenvalues, const Tensor& eigenvectors, const Tensor& infos, bool upper, bool compute_eigenvectors) {
+  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES(eigenvectors.scalar_type(), "linalg_eigh_cuda", [&] {
+    apply_syevj_batched<scalar_t>(eigenvalues, eigenvectors, infos, upper, compute_eigenvectors);
+  });
+}
 
+void linalg_eigh_cusolver(const Tensor& eigenvalues, const Tensor& eigenvectors, const Tensor& infos, bool upper, bool compute_eigenvectors) {
   // syevj is better than syevd for float32 dtype and matrix sizes 32x32 - 512x512
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/BatchLinearAlgebraLib.cu:1252]: this comment is misplaced (it refers to line 1256 now). Can you also add another comment with link to the data about why we are using syevjBatched for heuristics in line 1254?

**Comment by mruberry** [comment]
fyi @ngimel is merging this one

**Comment by xwang233** [comment]
Thank you!

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**